firestore/swift/firestore-smoketest/ViewController.swift

```diff
@@ -755,19 +755,39 @@ class ViewController: UIViewController {
         let docRef = db.collection("cities").document("BJ")
 
         docRef.getDocument { (document, error) in
+            if let error = error {
+                print("Error retrieving document: \(error)")
+                return
+            }
+
+            // Construct a Result type to encapsulate deserialization errors or
+            // successful deserialization. Note that if there is no error thrown
+            // the value may still be `nil`, indicating a successful deserialization
+            // of a value that does not exist.
+            //
+            // There are thus three cases to handle, which Swift lets us describe
+            // nicely with built-in sum types:
+            //
+            //      Result
+            //        /\
+            //   Error  Optional<City>
+            //               /\
+            //            Nil  City
             let result = Result {
-                try document.flatMap {
-                    try $0.data(as: City.self)
-                }
+              try document?.data(as: City.self)
             }
             switch result {
             case .success(let city):
                 if let city = city {
+                    // A `City` value was successfully initialized from the DocumentSnapshot.
                     print("City: \(city)")
                 } else {
+                    // A nil value was successfully initialized from the DocumentSnapshot,
```

Comment:
I don't think it's possible for there to be nil here since a Result success is not Optional.

Reply:
The type returned by the initialization call above is `Result<Optional<City>, Error>`. Normally a Result success is not optional, but it can be parameterized by any type, and in this case we've made it optional because of that success-but-nil case.

Reply:
It's a result of an optional value. That is, with the type spelled out, it's `Result<Optional<City>, Error>`.

Reply:
Yep, I just verified in a playground. Thanks!